### PR TITLE
Add MRI C Extension tests

### PR DESCRIPTION
**Pull request by bjfish**
This runs for suite or individual files:
```
jt test mri --cext
jt test mri cext/ruby/num2int/test_num2int.rb
```
So far the tests run are:
```
Bug::TestException#test_cleanup_data_error = 4.30 s = .
TestNum2int#test_fix2int = 0.05 s = .
TestNum2int#test_fix2long = 0.02 s = .
Test_BUG_5832#test_block_passing = 0.02 s = .
```

### CODE REVIEW

File: tool/jt.rb [hunk 1]
```diff
@@ -756,9 +765,31 @@ def test_mri(*args)
                       else
                         []
                       end
-
       files_to_run = (include_files - exclude_files)
+      
+    end
+
+    cext_tests = files_to_run.select { |f| f.include?("cext/ruby") }
```
Comment (eregon) [inline tool/jt.rb]: It seems better to set a variable above when args includes `--cext`, unless this should be triggered in other cases?

File: tool/jt.rb [hunk 2]
```diff
@@ -756,9 +765,31 @@ def test_mri(*args)
...
+                          parent_dir = match[1].split('/').slice(0..-2).join('/')
+                          "#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c/#{parent_dir}"
+                        end
+                      else
+                        "#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c/#{match[1]}"
+                      end
+        compile_cext("#{match[1].split("/")[1]}", compile_dir)
+      else
+        puts "c require not found for cext test: #{test_path}"
+      end
```
Comment (eregon) [inline tool/jt.rb]: Why is this here and not in run_mri_tests? To compile only once for `jt retag`?

File: tool/jt.rb [hunk 3]
```diff
@@ -756,9 +765,31 @@ def test_mri(*args)
...
                       end
-
       files_to_run = (include_files - exclude_files)
+      
+    end
+
+    cext_tests = files_to_run.select { |f| f.include?("cext/ruby") }
+    cext_tests.each do |test|
+      test_path = "#{TRUFFLERUBY_DIR}/test/mri/tests/#{test}"
+      match = File.open(test_path).read().match(/^require ['"]c\/(.*?)["']/)
```
Comment (eregon) [inline tool/jt.rb]: `File.read` to avoid keeping a FD opened.

File: tool/jt.rb [hunk 4]
```diff
@@ -756,9 +765,31 @@ def test_mri(*args)
...
+      if match
+        compile_dir = if match[1].include?('/')
+                        if Dir.exists?("#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c/#{match[1]}")
+                           "#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c/#{match[1]}"
+                        else
+                          parent_dir = match[1].split('/').slice(0..-2).join('/')
```
Comment (eregon) [inline tool/jt.rb]: Is this the same as File.dirname?

File: tool/jt.rb [hunk 5]
```diff
@@ -756,9 +765,31 @@ def test_mri(*args)
...
+                          "#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c/#{parent_dir}"
+                        end
+                      else
+                        "#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c/#{match[1]}"
```
Comment (eregon) [inline tool/jt.rb]: Can you extract `"#{TRUFFLERUBY_DIR}/test/mri/tests/cext/c` in a local or constant?

File: tool/jt.rb (hunk 1 above)
Comment (bjfish) [inline tool/jt.rb]: I was to handle a single test without `--cext`: 

`jt test mri cext/ruby/num2int/test_num2int.rb`

Comment (eregon) [inline tool/jt.rb]: I see, makes sense then.

File: lib/cext/ruby.h [hunk 6]
```diff
@@ -420,6 +420,19 @@ int FIX2INT(VALUE value);
...
 long FIX2LONG(VALUE value);
 unsigned long FIX2ULONG(VALUE value);
+
+short rb_num2short(VALUE);
+unsigned short rb_num2ushort(VALUE);
+
+#define RB_NUM2SHORT(x) rb_num2short(x)
+#define RB_NUM2USHORT(x) rb_num2ushort(x)
+#define NUM2SHORT(x) RB_NUM2SHORT(x)
+#define NUM2USHORT(x) RB_NUM2USHORT(x)
```
Comment (bjfish) [inline lib/cext/ruby.h]: Fixed.

File: tool/jt.rb [hunk 7]
```diff
@@ -642,18 +642,24 @@ def cextc(cext_dir, *clang_opts)
     compile_cext(name, ext_dir, target, *clang_opts)
   end
 
-  def compile_cext(name, ext_dir, target, *clang_opts)
+  def compile_cext(name, ext_dir, target = nil, *clang_opts)
```
Comment (bjfish) [inline tool/jt.rb]: Fixed.

File: tool/jt.rb [hunk 8]
```diff
@@ -642,18 +642,24 @@ def cextc(cext_dir, *clang_opts)
...
     extconf = "#{ext_dir}/extconf.rb"
     raise "#{extconf} does not exist" unless File.exist?(extconf)
 
+    args = if ext_dir.include?("test/mri/tests/cext/c")
+             ['-rmkmf']
+           else
+             []
+           end
```
Comment (bjfish) [inline tool/jt.rb]: Fixed.

File: tool/jt.rb (hunk 3 above)
Comment (bjfish) [inline tool/jt.rb]: Fixed.

File: tool/jt.rb (hunk 4 above)
Comment (bjfish) [inline tool/jt.rb]: Yes, fixed.

File: tool/jt.rb (hunk 5 above)
Comment (bjfish) [inline tool/jt.rb]: Yes, Fixed.

File: tool/jt.rb (hunk 2 above)
Comment (bjfish) [inline tool/jt.rb]: I've moved this to `run_mri_tests`.

File: lib/cext/ruby.h [hunk 9]
```diff
@@ -420,6 +420,16 @@ int FIX2INT(VALUE value);
 unsigned int FIX2UINT(VALUE value);
 long FIX2LONG(VALUE value);
 unsigned long FIX2ULONG(VALUE value);
+
+short rb_num2short(VALUE);
+unsigned short rb_num2ushort(VALUE);
```
Comment (eregon) [inline lib/cext/ruby.h]: Functions should have their arguments names in the header as well (see above)

File: tool/jt.rb [hunk 10]
```diff
@@ -776,7 +780,28 @@ def run_mri_tests(extra_args, test_files, run_options = {})
...
     }
 
+    cext_tests = test_files.select { |f| f.include?("cext/ruby") }
+    cext_tests.each do |test|
+      test_path = "#{TRUFFLERUBY_DIR}/test/mri/tests/#{test}"
+      match = File.read(test_path).match(/^require ['"]c\/(.*?)["']/)
+      if match
+        compile_dir = if match[1].include?('/')
+                        if Dir.exists?("#{MRI_TEST_CEXT_DIR}/#{match[1]}")
+                           "#{MRI_TEST_CEXT_DIR}/#{match[1]}"
```
Comment (eregon) [inline tool/jt.rb]: Wrong indentation

File: lib/cext/ruby.h (hunk 9 above)
Comment (bjfish) [inline lib/cext/ruby.h]: Fixed.

File: tool/jt.rb (hunk 10 above)
Comment (bjfish) [inline tool/jt.rb]: Fixed.
